File: DiabloCharacterTracker.Server/Services/CharacterSkillServices/CharacterSkillService.cs
using DiabloCharacterTracker.Server.Data;
using DiabloCharacterTracker.Server.DTOs;
using DiabloCharacterTracker.Server.Requests.AddRequests;
using Microsoft.EntityFrameworkCore;
using Microsoft.EntityFrameworkCore.Internal;

namespace DiabloCharacterTracker.Server.Services.CharacterSkillServices;

public class CharacterSkillService : ICharacterSkillService
{
    private readonly IDbContextFactory<DiabloDbContext> dbContextFactory;
    public CharacterSkillService(IDbContextFactory<DiabloDbContext> dbContextFactory)
    {
        this.dbContextFactory = dbContextFactory;
    }
    public async Task<int> AddCharacterSkill(AddCharacterSkillRequest request)
    {
        using var context = await dbContextFactory.CreateDbContextAsync();

        var possibleCharacterSkill = await context.CharacterSkills
            .Where(x => x.PlayableCharacterId == request.PlayableCharacterId)
            .Where(x => x.SkillId == request.SkillId)
            .FirstOrDefaultAsync();

        if (possibleCharacterSkill != null)
        {
            return 0;
        }

        CharacterSkill newCharackterSkill = new CharacterSkill()
        {
            PlayableCharacterId = request.PlayableCharacterId,
            SkillId = request.SkillId,
        };

        if (newCharackterSkill == null)
        {
            return 0;
        }

        await context.AddAsync(newCharackterSkill);
        await context.SaveChangesAsync();
        return newCharackterSkill.Id;
    }

    public async Task<List<SkillDTO>> GetAllSkillsForCharacter(int characterId)
    {
        using var context = await dbContextFactory.CreateDbContextAsync();

        var characterSkills = await context.CharacterSkills
            .Include(x => x.Skill)
            .Where(x => x.PlayableCharacterId == characterId)
            .ToListAsync();

        if (characterSkills == null)
        {
            return new List<SkillDTO>();
        }

        return characterSkills.Select(x => x.Skill!.ToDTO()).ToList();
    }

    public async Task<bool> RemoveCharacterSkill(int characterSkillId)
    {
        using var context = await dbContextFactory.CreateDbContextAsync();

        var characterSkillToRemove = await context.CharacterSkills
            .Where(x => x.Id == characterSkillId)
            .FirstOrDefaultAsync();

        if (characterSkillToRemove == null)
        {
            return false;
        }

        context.Remove(characterSkillToRemove);
        await context.SaveChangesAsync();
        return true;
    }
}

File: diablocharactertracker.client/src/ApiServices/CharacterSkillService.ts
import { Skill } from "../Data/DTOs/skill";
import { AddCharacterSkillRequest } from "../Data/Requests/AddRequests/AddCharacterSkillRequest";
import axios from "axios";

export const CharacterSkillService = {
  AddCharacterSkill: async (
    addCharacterSkillRequest: AddCharacterSkillRequest
  ) => {
    try {
      const response = await axios.post<number>(
        `${import.meta.env.VITE_URL}/CharacterSkill/addcharacterskill`,
        addCharacterSkillRequest,
        {
          headers: {
            "Content-Type": "application/json",
          },
        }
      );
      return response.data;
    } catch (error) {
      console.error("Failed to add character skill");
      throw error;
    }
  },
  GetAllSkillsForCharacter: async (characterId: number) => {
    try {
      const response = await axios.get<Skill[]>(
        `${import.meta.env.VITE_URL}/CharacterSkill/getallskillsforcharacter`,
        {
          params: {
            characterId,
          },
        }
      );
      return response.data;
    } catch (error) {
      console.error("Failed to get all skills for character");
      throw error;
    }
  },
  RemoveCharacterSkill: async (characterSkillId: number) => {
    try {
      const response = await axios.delete<boolean>(
        `${import.meta.env.VITE_URL}/CharacterSkill/removecharacterskill`,
        {
          params: {
            characterSkillId,
          },
        }
      );
      return response.data;
    } catch (error) {
      console.error("Failed to remove character skill");
      throw error;
    }
  },
};